Sim/SimDelphesInterface/src/DelphesSimulation.cpp

```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+                           relation.Sim().Core().P4.Py*relation.Sim().Core().P4.Py +
+                           relation.Sim().Core().P4.Pz*relation.Sim().Core().P4.Pz +
+                           relation.Sim().Core().P4.Mass*relation.Sim().Core().P4.Mass);
+
+      totSimE += simE;
+      std::cout << " RefId: "    << setw(3)  << id+1
+                << " Rel E: "    << setw(10) << totSimE << " " << simE << " <-> " << recE;
+      if      (colMCParticles->at(id).Core().Type ==22) std::cout << " Gamma";
+      else if (colMCParticles->at(id).Core().Charge==0) std::cout << " Neutral";
+      std::cout << std::endl;*/
+    }
+
+    // Test print
+    /*std::cout << std::endl;*/
 
+    if (barePart!=nullptr) delete barePart;
+  } // For - towers
+}
+
+void DelphesSimulation::ConvertJets(const TObjArray* Input,
+                                    fcc::GenJetCollection*     colJets,
+                                    fcc::MCParticleCollection* colMCParticles,
+                                    fcc::GenJetParticleAssociationCollection* ascColJetsToMC)
+{
+  Candidate* cand = nullptr;
+  for(int j = 0; j < Input->GetEntries(); ++j) {
+      
     cand = static_cast<Candidate *>(Input->At(j));
-    fcc::Particle outptc = coll->create();
-    fcc::BareParticle& core = outptc.Core();
-    core.P4.Px         = (double) cand->Momentum.X();
-    core.P4.Py         = (double) cand->Momentum.Y();
+
+    auto jet      = colJets->create();
+    auto relation = ascColJetsToMC->create();
+
+    auto bareJet      = new fcc::BareJet();
+    bareJet->Area     = -1;
+    bareJet->P4.Px    = cand->Momentum.Px();
+    bareJet->P4.Py    = cand->Momentum.Py();
+    bareJet->P4.Pz    = cand->Momentum.Pz();
+    bareJet->P4.Mass  = cand->Mass;
+    jet.Core(*bareJet);
+
+    // Test print
+    /*double energy = sqrt(jet.Core().P4.Px*jet.Core().P4.Px +
+                         jet.Core().P4.Py*jet.Core().P4.Py +
+                         jet.Core().P4.Pz*jet.Core().P4.Pz +
+                         jet.Core().P4.Mass*jet.Core().P4.Mass);
+    std::cout << "Delphes Jet: "
+              << " Id: "       << setw(3)  << j+1
+              << " Px: "       << setw(10) << jet.Core().P4.Px
+              << " Py: "       << setw(10) << jet.Core().P4.Py
+              << " Pz: "       << setw(10) << jet.Core().P4.Pz
+              << " E: "        << setw(10) << energy
+              << " M: "        << setw(10) << jet.Core().P4.Mass << std::endl;*/
+
+    // Reference to MC - Delphes holds references to all objects related to the Jet object,
+    // several relations might exist -> find "recursively" in a tree history a MC particle
+    std::set<int> idRefMCPart; // Avoid double counting when referencingh MC particles
+
+    Candidate* refCand  = nullptr;
+    Candidate* ref2Cand = nullptr;
+    Candidate* ref3Cand = nullptr;
+    Candidate* ref4Cand = nullptr;
+
+    TIter it1(cand->GetCandidates());
```

Comment:
No, the Delphes code works until this level of deepness, but one can do it recursively (so something similar what Alice did). In some cases there is direct link, in some cases in goes through related particle, etc. I will think of improving the code.